src/ol/render/webgl/VectorStyleRenderer.js

```diff
@@ -166,6 +178,31 @@ class VectorStyleRenderer {
       );
     }
 
+    /**
+     * @type {function(import('../../Feature.js').FeatureLike): boolean}
+     * @private
+     */
+    this.featureFilter_ = null;
+    if (filter) {
+      const parsingContext = newParsingContext();
+      const compiled = buildExpression(filter, BooleanType, parsingContext);
+      this.featureFilter_ = (feature) => {
+        const evalContext = newEvaluationContext();
```

Comment:
This could be created once and then reused - as long as all properties are updated - to avoid a bit of unnecessary garbage creation.